src/app/reporting/ReportingEngine.cpp

```diff
@@ -0,0 +1,173 @@
+/*
+ *
+ *    Copyright (c) 2021 Project CHIP Authors
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+/**
+ *    @file
+ *      This file implements reporting engine for CHIP
+ *      Data Model profile.
+ *
+ */
+
+#include <app/InteractionModelEngine.h>
+#include <app/reporting/ReportingEngine.h>
+
+namespace chip {
+namespace app {
+namespace reporting {
+CHIP_ERROR ReportingEngine::Init()
+{
+    CHIP_ERROR err       = CHIP_NO_ERROR;
+    mMoreChunkedMessages = false;
+    mNumReportsInFlight  = 0;
+    mCurReadHandlerIdx   = 0;
+    return err;
+}
+
+CHIP_ERROR ReportingEngine::BuildAndSendSingleReportData(ReadHandler * apReadHandler)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+    chip::System::PacketBufferTLVWriter reportDataWriter;
+    ReportData::Builder reportDataBuilder;
+    chip::System::PacketBufferHandle bufHandle = System::PacketBufferHandle::New(chip::app::kMaxSecureSduLengthBytes);
+
+    VerifyOrExit(!bufHandle.IsNull(), err = CHIP_ERROR_NO_MEMORY);
+
+    reportDataWriter.Init(std::move(bufHandle));
+
+    // Create a report data.
+    err = reportDataBuilder.Init(&reportDataWriter);
+    SuccessOrExit(err);
+
+    // TODO: Fill in the EventList.
+    // err = BuildSingleReportDataEventList(reportDataBuilder, apReadHandler);
+    // SuccessOrExit(err);
+
+    // TODO: Add mechanism to set mSuppressResponse to handle status reports for multiple reports
+    // TODO: Add more chunk message support, currently mMoreChunkedMessages is always false.
+    if (mMoreChunkedMessages)
+    {
+        reportDataBuilder.MoreChunkedMessages(mMoreChunkedMessages);
+    }
+
+    reportDataBuilder.EndOfReportData();
+    SuccessOrExit(reportDataBuilder.GetError());
+
+    err = reportDataWriter.Finalize(&bufHandle);
```

Comment:
you std::move'd the bufHandle above. I do not think you are allowed to reference it anymore.

Reply:
Thanks for comments.
Finalize would get the buffer handle back from PacketBufferTLVWriter,  CHIP_ERROR Finalize(chip::System::PacketBufferHandle * outBuffer)
    {
        CHIP_ERROR err = Finalize();
        *outBuffer     = mBackingStore.Release();
        return err;
    },
basically
reportDataWriter.Init(std::move(bufHandle));
err = reportDataWriter.Finalize(&bufHandle);
then i get the bufHandle back per the usage in PacketBufferTLVWriter